In [34]:
import numpy as np

In [35]:
# Task 1: Dataset Creation and Understanding
data = np.array([
    [501,45,3,120,60],
    [502,30,2,80,50],
    [503,60,4,np.nan,70],
    [504,25,1,60,40],
    [505,80,5,200,90],
    [506,-20,2,70,50],
    [507,35,2,90,55],
    [508,50,3,np.nan,65],
    [509,90,6,250,100],
    [510,200,4,150,80],
    [511,55,3,110,60],
    [512,40,2,95,55]
])

In [36]:
print("shape:", data.shape)
print("rows:", data.shape[0])
print("columns:", data.shape[1])

shape: (12, 5)
rows: 12
columns: 5


player_id-Unique player identifier
session_duration-Time spent in game
levels_completed-Number of levels completed
in_game_actions-Actions performed during gameplay
energy_used-Energy consumed

In [37]:
#Task 2 – Indexing & Slicing
#Extract columns
player_ids = data[:,0]
session_duration = data[:,1]
levels_completed = data[:,2]
print("Player IDs")
print(player_ids)
print("Session Duration")
print(session_duration)
print("Levels Completed")
print(levels_completed)

Player IDs
[501. 502. 503. 504. 505. 506. 507. 508. 509. 510. 511. 512.]
Session Duration
[ 45.  30.  60.  25.  80. -20.  35.  50.  90. 200.  55.  40.]
Levels Completed
[3. 2. 4. 1. 5. 2. 2. 3. 6. 4. 3. 2.]


In [38]:
#First 5 players
firstfive = data[:5]
print(firstfive)

[[501.  45.   3. 120.  60.]
 [502.  30.   2.  80.  50.]
 [503.  60.   4.  nan  70.]
 [504.  25.   1.  60.  40.]
 [505.  80.   5. 200.  90.]]


In [39]:
# Access one player
player509 = data[4]
print(player509)

[505.  80.   5. 200.  90.]


Through numpy slicing we can get quick access to subsets. Here teh entire columns are extracted efficiently without using loops.

In [40]:
#Task 3 – Basic Analysis
avg_dur = np.nanmean(data[:,1])
print("Average Duration:", avg_dur)

Average Duration: 57.5


In [41]:
max_dur = np.nanmax(data[:,1])
min_dur = np.nanmin(data[:,1])
print("Maximum Duration:", max_dur)
print("Minimum Duration:", min_dur)

Maximum Duration: 200.0
Minimum Duration: -20.0


In [42]:
avg_levels = np.nanmean(data[:,2])
print("Average Levels:", avg_levels)

Average Levels: 3.0833333333333335


In [43]:
total_actions = np.nansum(data[:,3])
print("Total Actions:", total_actions)

Total Actions: 1225.0


We can conclude that in average most players spend 57.5 minutes per session. The longest session was 200 minutes. The existence of negative duration indicates that the data is not perfect.The overall engagement was very strong. For some players the number of actions are numerous compared to their shorter sessions.

In [44]:
#Task 4 – Vectorization
# Bonus actions
bonusactions = data[:,3] + 10
print(bonusactions)

[130.  90.  nan  70. 210.  80. 100.  nan 260. 160. 120. 105.]


In [45]:
# High performers
high_perform = data[data[:,2] > 4]
print(high_perform)

[[505.  80.   5. 200.  90.]
 [509.  90.   6. 250. 100.]]


In [46]:
# Low activity players
lowactivity = data[data[:,3] < 80]
print(lowactivity)

[[504.  25.   1.  60.  40.]
 [506. -20.   2.  70.  50.]]


The highest performing players are found to be players 505 and 509 and have completed more than 5 levels.Players that have less than 80 actions show lower engagements.

In [47]:
#Task 5 – Feature Engineering
performancescore = data[:,2] / data[:,1]
dataextended = np.column_stack((data, performancescore))

print(dataextended)

[[ 5.01000000e+02  4.50000000e+01  3.00000000e+00  1.20000000e+02
   6.00000000e+01  6.66666667e-02]
 [ 5.02000000e+02  3.00000000e+01  2.00000000e+00  8.00000000e+01
   5.00000000e+01  6.66666667e-02]
 [ 5.03000000e+02  6.00000000e+01  4.00000000e+00             nan
   7.00000000e+01  6.66666667e-02]
 [ 5.04000000e+02  2.50000000e+01  1.00000000e+00  6.00000000e+01
   4.00000000e+01  4.00000000e-02]
 [ 5.05000000e+02  8.00000000e+01  5.00000000e+00  2.00000000e+02
   9.00000000e+01  6.25000000e-02]
 [ 5.06000000e+02 -2.00000000e+01  2.00000000e+00  7.00000000e+01
   5.00000000e+01 -1.00000000e-01]
 [ 5.07000000e+02  3.50000000e+01  2.00000000e+00  9.00000000e+01
   5.50000000e+01  5.71428571e-02]
 [ 5.08000000e+02  5.00000000e+01  3.00000000e+00             nan
   6.50000000e+01  6.00000000e-02]
 [ 5.09000000e+02  9.00000000e+01  6.00000000e+00  2.50000000e+02
   1.00000000e+02  6.66666667e-02]
 [ 5.10000000e+02  2.00000000e+02  4.00000000e+00  1.50000000e+02
   8.00000000e+01  2.0000

In [48]:
#Step 1: Handle Missing Values
cleaneddata = data.copy()
actions = cleaneddata[:,3]
actionmean = np.nanmean(actions)
cleaneddata[np.isnan(actions),3] = actionmean
print(cleaneddata)

[[501.   45.    3.  120.   60. ]
 [502.   30.    2.   80.   50. ]
 [503.   60.    4.  122.5  70. ]
 [504.   25.    1.   60.   40. ]
 [505.   80.    5.  200.   90. ]
 [506.  -20.    2.   70.   50. ]
 [507.   35.    2.   90.   55. ]
 [508.   50.    3.  122.5  65. ]
 [509.   90.    6.  250.  100. ]
 [510.  200.    4.  150.   80. ]
 [511.   55.    3.  110.   60. ]
 [512.   40.    2.   95.   55. ]]


In [49]:
durationmean = np.mean(
    cleaned_data[cleaned_data[:,1] > 0,1]
)
cleaned_data[cleaned_data[:,1] < 0,1] = duration_mean
print(cleaneddata)

[[501.   45.    3.  120.   60. ]
 [502.   30.    2.   80.   50. ]
 [503.   60.    4.  122.5  70. ]
 [504.   25.    1.   60.   40. ]
 [505.   80.    5.  200.   90. ]
 [506.  -20.    2.   70.   50. ]
 [507.   35.    2.   90.   55. ]
 [508.   50.    3.  122.5  65. ]
 [509.   90.    6.  250.  100. ]
 [510.  200.    4.  150.   80. ]
 [511.   55.    3.  110.   60. ]
 [512.   40.    2.   95.   55. ]]


In [50]:
duration = cleaneddata[:,1]
mean_duration = np.mean(duration)
std_duration = np.std(duration)
upper_limit = mean_duration + 2 * std_duration
lower_limit = mean_duration - 2 * std_duration
print("Upper Limit:", upper_limit)
print("Lower Limit:", lower_limit)

Upper Limit: 158.70111988840176
Lower Limit: -43.70111988840176


In [51]:
outliers = (duration > upper_limit) | (duration < lower_limit)
print("Outlier Positions")
print(np.where(outliers))

Outlier Positions
(array([9]),)


In [52]:
cleaneddata[outliers,1] = mean_duration
print(cleaneddata)

[[501.   45.    3.  120.   60. ]
 [502.   30.    2.   80.   50. ]
 [503.   60.    4.  122.5  70. ]
 [504.   25.    1.   60.   40. ]
 [505.   80.    5.  200.   90. ]
 [506.  -20.    2.   70.   50. ]
 [507.   35.    2.   90.   55. ]
 [508.   50.    3.  122.5  65. ]
 [509.   90.    6.  250.  100. ]
 [510.   57.5   4.  150.   80. ]
 [511.   55.    3.  110.   60. ]
 [512.   40.    2.   95.   55. ]]


In [53]:
#Task 7 – Final Insights
avgduration_clean = np.mean(cleaneddata[:,1])
avglevels_clean = np.mean(cleaneddata[:,2])
avgactions_clean = np.mean(cleaneddata[:,3])
print(avgduration_clean)
print(avglevels_clean)
print(avgactions_clean)

45.625
3.0833333333333335
122.5


In [54]:
correlation = np.corrcoef(
    cleaneddata[:,1],
    cleaneddata[:,2]
)
print(correlation)

[[1.         0.82039082]
 [0.82039082 1.        ]]


In [55]:
sorted_players = cleaneddata[
    np.argsort(cleaneddata[:,2])[::-1]
]
print(sorted_players[:3])

[[509.   90.    6.  250.  100. ]
 [505.   80.    5.  200.   90. ]
 [503.   60.    4.  122.5  70. ]]


In [56]:
print("Median Duration:",
      np.median(cleaneddata[:,1]))
print("Median Actions:",
      np.median(cleaneddata[:,3]))

Median Duration: 47.5
Median Actions: 115.0


Players 509 and 505 are strong performers with 509 being slightly better by completing 6 levels and 250 actions. The dataset was cleaned by handling missing values,negative values and outliers which could haveaffected the indights obtained.An average player completes 2-4 levels per session. longer sessions is direclty proportional to levels completed in most cases..